pkg/runtime/grpc/grpc-runtime.go

```diff
@@ -0,0 +1,440 @@
+		return nil, fmt.Errorf("failed to creating RESTConfig: %w", err)
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set up trace client: %w", err)
+	}
+
+	opts := metav1.ListOptions{LabelSelector: "k8s-app=gadget"}
+	pods, err := client.CoreV1().Pods("gadget").List(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("getting pods: %w", err)
+	}
+
+	if len(pods.Items) == 0 {
+		return nil, fmt.Errorf("no gadget pods found. Is Inspektor Gadget deployed?")
+	}
+
+	if len(nodes) == 0 {
+		res := make([]gadgetPod, 0, len(pods.Items))
+
+		for _, pod := range pods.Items {
+			res = append(res, gadgetPod{name: pod.Name, node: pod.Spec.NodeName})
+		}
+
+		return res, nil
+	}
+
+	// Remove possible duplicates in nodes
+	resMap := map[string]string{}
+	for _, node := range nodes {
+		resMap[node] = ""
+	}
+
+	// Filter nodes
+	for node := range resMap {
+		for _, pod := range pods.Items {
+			if node == pod.Spec.NodeName {
+				resMap[node] = pod.Name
+				break
+			}
+		}
+	}
+
+	res := make([]gadgetPod, 0, len(nodes))
+	for node, pod := range resMap {
+		if pod == "" {
+			log.Warnf("node %q not found or without gadget pod", node)
+			continue
+		}
+		res = append(res, gadgetPod{name: pod, node: node})
+	}
+
+	return res, nil
+}
+
+func (r *Runtime) RunGadget(gadgetCtx runtime.GadgetContext) (map[string][]byte, error) {
+	// Get nodes to run on
+	nodes := gadgetCtx.RuntimeParams().Get(ParamNode).AsStringSlice()
+	pods, err := r.getGadgetPods(gadgetCtx.Context(), nodes)
+	if err != nil {
+		return nil, fmt.Errorf("get gadget pods: %w", err)
+	}
+	if len(pods) == 0 {
+		return nil, fmt.Errorf("no nodes found to run on")
+	}
+
+	if gadgetCtx.GadgetDesc().Type() == gadgets.TypeTraceIntervals {
+		gadgetCtx.Parser().EnableSnapshots(
+			gadgetCtx.Context(),
+			time.Duration(gadgetCtx.GadgetParams().Get(gadgets.ParamInterval).AsInt32())*time.Second,
+			2,
+		)
+	}
+
+	results := make(map[string][]byte)
+	var resultsLock sync.Mutex
+
+	wg := sync.WaitGroup{}
+	for _, pod := range pods {
+		wg.Add(1)
+		go func(pod gadgetPod) {
+			gadgetCtx.Logger().Debugf("running gadget on node %q", pod.node)
+			res, err := r.runGadget(gadgetCtx, pod)
+			if err != nil {
+				gadgetCtx.Logger().Errorf("node %q: %w", pod.node, err)
+			}
+			if res != nil {
+				resultsLock.Lock()
+				results[pod.node] = res
+				resultsLock.Unlock()
+			}
+			wg.Done()
+		}(pod)
+	}
+
+	wg.Wait()
+	return results, nil
+}
+
+func (r *Runtime) runGadget(gadgetCtx runtime.GadgetContext, pod gadgetPod) ([]byte, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	dialOpt := grpc.WithContextDialer(func(ctx context.Context, s string) (net.Conn, error) {
+		return NewK8SExecConn(ctx, pod, time.Second*30)
+		// return NewK8SPortForwardConn(ctx, s, time.Second*30)
+	})
+
+	conn, err := grpc.DialContext(ctx, "", dialOpt, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
```

Comment:
Certificates, for example. It's not used here, because we use a unix socket that we don't expose (therefore essentially trusting the k8s authorization).